### PR TITLE
Feature: Oracle outer join operator (+) support

### LINKED ISSUE #685: Oracle special outer join syntax not supported
## Enhancement
Oracle has a special syntax for outer join that is not supported in ivorySQL.

The syntax looks like this:
```
SELECT o.orderno, c.customername
FROM orders o, customer c
WHERE o.customerid = c.customerid (+);
```

This command is exactly the same as:
```
SELECT o.orderno, c.customername
FROM orders o
LEFT OUTER JOIN customer c
ON o.customerid = c.customerid;
```

It is possible to continue outer joins on more tables.

This is very common syntax in old Oracle applications and this is enhancement needed in order to migrate to ivorySQL for many such systems.

Ps. Thanks for your excellent work :-)

### PR DESCRIPTION
**Pull request by yasir-hussain-shah**
This commit adds support for Oracle outer join operator (+).

Implementation details:
- liboracle_parser.c: Recognize "(+)" sequence as ORAJOINOPR token
- ora_gram.y: Add parser rules for column references with (+)
- parsenodes.h: - ColumnRef structure: Add ora_join_op_exists flag to mark join columns - SelectStmt structure: Add ora_join_op_exists flag for query-level detection
- parse_clause.c: Implement transformOraJoinClauses() to transform WHERE conditions with (+) into ANSI JOIN syntax
- parse_expr.c: Capture column references during transformation
- parse_node.h: Add OraJoinState for tracking join context

Test coverage:
- Basic left and right outer joins
- Filtered joins with WHERE conditions
- Complex conditions with AND/OR
- Aggregation with outer joins
- Multiple outer joins in single query
- Edge cases and error conditions

fix #685 

## Summary by CodeRabbit

* **New Features**
  * Added support for Oracle-style outer joins using the `(+)` operator.
  * Supports left and right joins, filtering, aggregation, multiple joins, self-joins, and complex conditions.
  * Supports joins involving subqueries, common table expressions, values lists, table functions, and qualified column references.
  * Supports spaced or commented `(+)` operators.

* **Bug Fixes**
  * Improved handling of join filters and nested expressions.
  * Added clearer errors for unsupported or invalid `(+)` usage.

* **Tests**
  * Added comprehensive regression coverage for common and edge-case Oracle outer-join queries.

### CODE REVIEW

File: src/backend/oracle_parser/liboracle_parser.c [hunk 1]
```diff
@@ -865,6 +868,33 @@ ora_base_yylex(YYSTYPE *lvalp, YYLTYPE *llocp, ora_core_yyscan_t yyscanner)
...
+						yyextra->loc_pushback++;
+
+						yyextra->have_lookahead = true;
+						return cur_token;
+					}
+
+					push_back_token(yyscanner, next_token, &aux1);
+				}
+			}
+			break;
```
[the remark this answers is not in the document]
Comment (yasir-hussain-shah) [reply src/backend/oracle_parser/liboracle_parser.c:910]: Fixed.

File: src/backend/oracle_parser/ora_gram.y [hunk 2]
```diff
@@ -20491,6 +20493,16 @@ columnref:	ColId
...
+			| ColId ORAJOINOPR
+				{
+					$$ = makeColumnRef($1, NIL, @1, yyscanner);
+					((ColumnRef *) $$)->ora_join_op_exists = true;
+				}
+			| ColId indirection ORAJOINOPR
+				{
+					$$ = makeColumnRef($1, $2, @1, yyscanner);
+					((ColumnRef *) $$)->ora_join_op_exists = true;
+				}
```
[the remark this answers is not in the document]
Comment (yasir-hussain-shah) [reply src/backend/oracle_parser/ora_gram.y:20522]: Fixed.

File: src/backend/parser/parse_clause.c [hunk 3]
```diff
@@ -1997,6 +1998,241 @@ setNamespaceLateralState(List *namespace, bool lateral_only, bool lateral_ok)
...
+
+				rv = makeRangeVar(NULL, get_rel_name(lrte->relid), -1);
+				n->larg = (Node *)rv;
+				rv->inh = true;
+				rv->alias = makeAlias(lrte->eref->aliasname, NIL);
+
+				rv = makeRangeVar(NULL, get_rel_name(rrte->relid), -1);
+				n->rarg = (Node *)rv;
+				rv->inh = true;
+				rv->alias = makeAlias(rrte->eref->aliasname, NIL);
```
[the remark this answers is not in the document]
Comment (yasir-hussain-shah) [reply src/backend/parser/parse_clause.c]: Fixed.

File: src/backend/parser/parse_clause.c [hunk 4]
```diff
@@ -1997,6 +1998,241 @@ setNamespaceLateralState(List *namespace, bool lateral_only, bool lateral_ok)
...
+					JoinExpr *prev = linitial(*joinClause);
+					n->larg = (Node *)prev;
+
+					*joinClause = list_delete_first(*joinClause);
+					*joinClause = lappend(*joinClause, (Node *)n);
+				}
+				else
+				{
+					*joinClause = lappend(*joinClause, n);
+				}
```
[the remark this answers is not in the document]
Comment (yasir-hussain-shah) [reply src/backend/parser/parse_clause.c]: Fixed.

File: src/backend/parser/parse_expr.c [hunk 5]
```diff
@@ -739,6 +739,14 @@ transformColumnRefInternal(ParseState *pstate, ColumnRef *cref, bool missing_ok)
 						node = transformWholeRowRef(pstate, nsitem, levels_up,
 													cref->location);
 				}
+
+				if (pstate->p_orajoin_state != NULL)
+				{
+					OraJoinState *ojs = (OraJoinState *) pstate->p_orajoin_state;
+					ojs->var = (Var *) node;
+					ojs->cref = cref;
+				}
```
Comment (yasir-hussain-shah) [inline src/backend/parser/parse_expr.c]: Fixed.

File: src/oracle_test/regress/expected/oracle_join_operator.out [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+INSERT INTO t1 VALUES (30, 't1_c');
+INSERT INTO t1 VALUES (40, 't1_d');
+INSERT INTO t2 VALUES (1, 't2_a', 10);
+INSERT INTO t2 VALUES (2, 't2_b', 10);
+INSERT INTO t2 VALUES (3, 't2_c', 20);
+INSERT INTO t2 VALUES (4, 't2_d', NULL);
+INSERT INTO t2 VALUES (5, 't2_e', 50);
+INSERT INTO t3 VALUES (1, 1, 't3_a');
+INSERT INTO t3 VALUES (2, 2, 't3_b');
+INSERT INTO t3 VALUES (3, 99, 't3_c');
```
[the remark this answers is not in the document]
Comment (yasir-hussain-shah) [reply src/oracle_test/regress/expected/oracle_join_operator.out:31]: There's no need make this change because make check and make oracle-check verify the results.

File: src/backend/parser/parse_clause.c [hunk 7]
```diff
@@ -1997,6 +1998,320 @@ setNamespaceLateralState(List *namespace, bool lateral_only, bool lateral_ok)
...
+	if (rtindex > 0 && rtindex <= build->num_rtindexes &&
+		build->rtindex_to_node[rtindex] != NULL)
+		return build->rtindex_to_node[rtindex];
+
+	node = get_orajoin_fromclause_node(build->fromClause, rte, rtindex);
+	if (node != NULL)
+		return node;
+
+	rv = makeRangeVar(NULL, get_rel_name(rte->relid), -1);
+	rv->inh = true;
```
[the remark this answers is not in the document]
Comment (yasir-hussain-shah) [reply src/backend/parser/parse_clause.c:2093]: Done.

File: src/backend/parser/parse_expr.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (yasir-hussain-shah) [reply src/backend/parser/parse_expr.c]: Done.

File: src/oracle_test/regress/sql/oracle_join_operator.sql [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+-- with both t2 rows whose c3=10, t1_b/t2_c, t1_c/NULL, t1_d/NULL) = 15
+-- rows total. What matters here is that the query parses and executes
+-- without error and without either pair leaking rows/aliases into the
+-- other -- not any specific row ordering within the 15.
+SELECT t4.label AS t4_label, t5.label AS t5_label,
+       t1.c2 AS t1_c2, t2.c2 AS t2_c2
+FROM t4, t5, t1, t2
+WHERE t4.id = t5.id(+)
+AND t1.c1 = t2.c3(+)
+ORDER BY t4.label, t1.c2;
```
[the remark this answers is not in the document]
Comment (yasir-hussain-shah) [reply src/oracle_test/regress/sql/oracle_join_operator.sql]: Done.

File: src/backend/parser/parse_clause.c [hunk 9]
```diff
@@ -1997,6 +1998,320 @@ setNamespaceLateralState(List *namespace, bool lateral_only, bool lateral_ok)
...
+
+				if (ljs.cref->ora_join_op_exists && rjs.cref->ora_join_op_exists)
+					elog(ERROR, "only one table can be used as outer join");
+
+				/* not a oracle join condition */
+				if (!ljs.cref->ora_join_op_exists && !rjs.cref->ora_join_op_exists)
+				{
+					*whereClause = lappend(*whereClause, expr);
+					return (Node *) aexpr;
+				}
```
[the remark this answers is not in the document]
Comment (yasir-hussain-shah) [reply src/backend/parser/parse_clause.c]: Done.

File: src/backend/parser/parse_clause.c [hunk 10]
```diff
@@ -1997,6 +1998,320 @@ setNamespaceLateralState(List *namespace, bool lateral_only, bool lateral_ok)
...
+get_orajoin_fromclause_node(List *fromClause, RangeTblEntry *rte, int rtindex)
+{
+	if (rte->rtekind == RTE_RELATION)
+		return NULL;			/* handled by the existing relid/RangeVar path */
+
+	if (fromClause == NIL || rtindex <= 0 || rtindex > list_length(fromClause))
+		return NULL;			/* out of range -- fall back to old behavior */
+
+	return copyObject((Node *) list_nth(fromClause, rtindex - 1));
+}
```
[the remark this answers is not in the document]
Comment (yasir-hussain-shah) [reply src/backend/parser/parse_clause.c:2070]: Recheck.

**Comment by hs-liuxh** [comment]
Thanks for your contribution. I will review it shortly.

**Comment by hs-liuxh** [comment]
Thank you for your work! The Oracle operator (+) is of immense value when migrating legacy Oracle code, and this PR is crucial for enhancing IvorySQL's Oracle compatibility.

I have tested the implementation and confirmed that the main workflow functions correctly. However, during edge-case testing (pasted at the bottom), I found a few behavioral inconsistencies with Oracle.  This may be because the recognition scope for the outer join operator (+) is too narrow, causing it to be silently ignored in certain scenarios.

These 4 test cases reuse the test data provided in the original PR. Could you please incorporate these test cases into your implementation, and ensure the output matches Oracle's behavior?

### Oracle
```sql
SQL> SELECT banner FROM v$version;                                         

BANNER
-------------------------------------------------------------------------------------------------------
Oracle AI Database 26ai Free Release 23.26.0.0.0 - Develop, Learn, and Run for Free

SQL> SELECT t2.c2,t1.c2 FROM t2,t1 WHERE t2.c3=t1.c1(+) AND t1.c2(+)='t1_a';

C2                                                 C2
-------------------------------------------------- --------------------------------------------------
t2_a                                               t1_a
t2_b                                               t1_a
t2_e                                               <NULL>
t2_d                                               <NULL>
t2_c                                               <NULL>

SQL> SELECT t2.c2,t1.c2 FROM t2,t1 WHERE t2.c3=t1.c1(+) AND t1.c2(+) IS NULL;

C2                                                 C2
-------------------------------------------------- --------------------------------------------------
t2_e                                               <NULL>
t2_a                                               <NULL>
t2_b                                               <NULL>
t2_d                                               <NULL>
t2_c                                               <NULL>

SQL> SELECT t2.c2,t1.c2 FROM t2,t1 WHERE t2.c3=t1.c1(+)+0;

C2                                                 C2
-------------------------------------------------- --------------------------------------------------
t2_a                                               t1_a
t2_b                                               t1_a
t2_c                                               t1_b
t2_e                                               <NULL>
t2_d                                               <NULL>

SQL> SELECT t2.c2,t1.c2 FROM t2,t1 WHERE t2.c2=upper(t1.c2(+));

C2                                                 C2
-------------------------------------------------- --------------------------------------------------
t2_c                                               <NULL>
t2_d                                               <NULL>
t2_e                                               <NULL>
t2_a                                               <NULL>
t2_b                                               <NULL>
```

### IvorySQL
```sql
[ORA]postgres=# SELECT t2.c2,t1.c2 FROM t2,t1 WHERE t2.c3=t1.c1(+) AND t1.c2(+)='t1_a';
  c2  |  c2  
------+------
 t2_a | t1_a
 t2_b | t1_a
(2 rows)

[ORA]postgres=# SELECT t2.c2,t1.c2 FROM t2,t1 WHERE t2.c3=t1.c1(+) AND t1.c2(+) IS NULL;
  c2  | c2 
------+----
 t2_e | 
 t2_d | 
(2 rows)

[ORA]postgres=# SELECT t2.c2,t1.c2 FROM t2,t1 WHERE t2.c3=t1.c1(+)+0;
  c2  |  c2  
------+------
 t2_a | t1_a
 t2_b | t1_a
 t2_c | t1_b
(3 rows)

[ORA]postgres=# SELECT t2.c2,t1.c2 FROM t2,t1 WHERE t2.c2=upper(t1.c2(+));
 c2 | c2 
----+----
(0 rows)

```

**Comment by hs-liuxh** [comment]
The inconsistencies in error reporting between IvorySQL and Oracle can be mainly categorized into three types:

1. Both Oracle and IvorySQL throw errors, but the error messages differ.
The current error messages provide little diagnostic value to users and expose internal implementation details.

2. Oracle does not report an error, whereas IvorySQL does.
3. Oracle throws an error, but IvorySQL silently accepts it.

```sql
--- Oracle 

SQL> SELECT t3.c3,t1.c2,t2.c2 FROM t1 JOIN t2 ON t1.c1=t2.c3, t3 WHERE t2.c1=t3.c2(+);
SELECT t3.c3,t1.c2,t2.c2 FROM t1 JOIN t2 ON t1.c1=t2.c3, t3 WHERE t2.c1=t3.c2(+)
                                                                        *
ERROR at line 1:
ORA-25156: old style outer join (+) cannot be used with ANSI joins
Help: https://docs.oracle.com/error-help/db/ora-25156/


SQL> SELECT count(*) FROM (t1 CROSS JOIN t3), t2 WHERE t2.c3=t1.c1(+);
SELECT count(*) FROM (t1 CROSS JOIN t3), t2 WHERE t2.c3=t1.c1(+)
                                                        *
ERROR at line 1:
ORA-25156: old style outer join (+) cannot be used with ANSI joins
Help: https://docs.oracle.com/error-help/db/ora-25156/


SQL> SELECT count(*) FROM t2,t1 WHERE t1.c1(+) IN (10,20);

  COUNT(*)
----------
        10

SQL> SELECT count(*) FROM t2,t1 WHERE t2.c3=t1.c1(+) AND t1.c1(+) NOT BETWEEN 1 AND 5;

  COUNT(*)
----------
         5

SQL> SELECT t1.c2(+) FROM t1;
SELECT t1.c2(+) FROM t1
            *
ERROR at line 1:
ORA-30563: outer join operator (+) is not allowed here
Help: https://docs.oracle.com/error-help/db/ora-30563/
```

```sql
--- IvorySQL

[ORA]postgres=# SELECT t3.c3,t1.c2,t2.c2 FROM t1 JOIN t2 ON t1.c1=t2.c3, t3 WHERE t2.c1=t3.c2(+);
ERROR:  unrecognized node type: 66
[ORA]postgres=# SELECT count(*) FROM (t1 CROSS JOIN t3), t2 WHERE t2.c3=t1.c1(+);
ERROR:  table name "t2" specified more than once
[ORA]postgres=# SELECT count(*) FROM t2,t1 WHERE t1.c1(+) IN (10,20);
ERROR:  unrecognized node type: 1
[ORA]postgres=# SELECT count(*) FROM t2,t1 WHERE t2.c3=t1.c1(+) AND t1.c1(+) NOT BETWEEN 1 AND 5;
ERROR:  unrecognized node type: 1

[ORA]postgres=# SELECT t1.c2(+) FROM t1;
  c2  
------
 t1_a
 t1_b
 t1_c
 t1_d
(4 rows)
```

**Comment by hs-liuxh** [comment]
One more behavioral inconsistency, raised separately.

This is the mirror of the existing `Multiple outer joins` test case in the PR: the existing
case pairs one preserved table with several null-side tables, while this case pairs one
null-side table with several preserved tables.

```sql
WHERE t1.c1 = t2.c3(+) AND t3.c1 = t2.c1(+)
```

Same test data as the previous comment.

### Oracle
```sql
SQL> SELECT t1.c2, t2.c2, t3.c3 FROM t1,t2,t3 WHERE t1.c1=t2.c3(+) AND t3.c1=t2.c1(+) ORDER BY 1,2,3;

C2         C2         C3
---------- ---------- ----------
t1_a       t2_a       t3_a
t1_a       t2_b       t3_b
t1_a       <NULL>     t3_c
t1_b       t2_c       t3_c
t1_b       <NULL>     t3_a
t1_b       <NULL>     t3_b
t1_c       <NULL>     t3_a
t1_c       <NULL>     t3_b
t1_c       <NULL>     t3_c
t1_d       <NULL>     t3_a
t1_d       <NULL>     t3_b
t1_d       <NULL>     t3_c

12 rows selected.
```

### IvorySQL
```sql
[ORA]postgres=# SELECT t1.c2, t2.c2, t3.c3 FROM t1,t2,t3
                WHERE t1.c1=t2.c3(+) AND t3.c1=t2.c1(+) ORDER BY 1,2,3;
  c2  |  c2  |  c3
------+------+------
 t1_a | t2_a | t3_a
 t1_a | t2_b | t3_b
 t1_b | t2_c | t3_c
(3 rows)
```

Could you please include this case in the implementation and make the output match Oracle?

File: src/backend/parser/parse_clause.c [hunk 11]
```diff
@@ -1997,6 +1998,700 @@ setNamespaceLateralState(List *namespace, bool lateral_only, bool lateral_ok)
...
+
+						f->rtindex = marked->var->varno;
+						f->expr = (Node *) aexpr;
+						build->pending_filters = lappend(build->pending_filters, f);
+						return (Node *) aexpr;
+					}
+
+					/* nothing (+)-related resolved; leave the predicate be */
+					*whereClause = lappend(*whereClause, expr);
+					return (Node *) aexpr;
```
[the remark this answers is not in the document]
Comment (yasir-hussain-shah) [reply src/backend/parser/parse_clause.c:2354]: I deliberately made `t1.c1(+) + 0` and `abs(t1.c1(+))` succeed rather than error, that was the whole point of the earlier fix loosening the gate to `orajoin_marker_walker(aexpr->lexpr/rexpr, ...)`. The expected output for both was captured directly from a real Oracle instance (shown above), and it returns 5 rows, not ORA-.... Flipping this to an error, as the bot suggests, would revert that fix and break the two tests it was written to satisfy.

On the technical premise itself, the bot claims "no plain Var is resolved" for these buried cases, which is why it thinks the code falls through to the untouched-fallback at lines 2316–2318. That's not accurate for the current implementation: `record_orajoin_operand()` fires from inside `transformColumnRef()`, which `transformExpr()` reaches recursively regardless of how deeply the `ColumnRef` is nested (inside a FuncCall like `abs(...)`, or an `A_Expr like + 0)`. So `rjs.var/ljs.var` do get populated for these two queries, and execution takes the normal two-sided join-building path further down (`lindex = ljs.var->varno; rindex = rjs.var->varno; ...`), never touching the fallback at 2316–2318 at all. That fallback is reserved for genuinely different situations — e.g. `a (+)`-marked column buried inside a construct `transformExpr` doesn't recurse into via `transformColumnRef` at all (a subquery, say), which is a real but separate gap from the one this comment describes.

So: declining this suggested change.

**Comment by yasir-hussain-shah** [comment]
@hs-liuxh 
Thanks for highlighting all issues and output differences with Oracle you mentioned above. All of those has been fixed now and test cases added too. Please have a look with latest version.

**Comment by yasir-hussain-shah** [comment]
Converged all subsequent changes into one and polished.
